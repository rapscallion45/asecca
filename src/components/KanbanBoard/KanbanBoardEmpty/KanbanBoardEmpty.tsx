import { FC } from 'react';
import { Box, Typography } from '@mui/material';
import AddIcon from '@mui/icons-material/Add';
import EditOutlinedIcon from '@mui/icons-material/EditOutlined';
import { IKanbanBoard } from '@/lib/api/api-types';
import FormModal from '@/modals/FormModal/FormModal';
import KanbanBoardForm from '../KanbanBoardForm';

/**
 * Kanban Board Empty Props
 *
 * @author Carl Scrivener {@link https://github.com/rapscallion45 GitHub}
 * @since 0.0.1
 *
 * @typedef IKanbanBoardEmptyProps
 * @prop {string} type - add or edit type to show whether board has content
 * @prop {IKanbanBoard} currentData - passed board data
 */
interface IKanbanBoardEmptyProps {
  type: string;
  currentData?: IKanbanBoard;
}

/**
 * Kanban Board Empty
 *
 * Kanban board empty interface for when no Kanban board data is loaded
 *
 * @author Carl Scrivener {@link https://github.com/rapscallion45 GitHub}
 * @since 0.0.1
 *
 * @component
 * @param {IEmptyBoardProps} props - component props
 * @returns {FC} - empty board functional component
 */
const KanbanBoardEmpty: FC<IKanbanBoardEmptyProps> = (props) => {
  const { type, currentData } = props;

  return (
    <div className=" bg-white dark:bg-[#2b2c37] h-screen w-screen flex flex-col  items-center justify-center">
      <Typography variant={type === 'edit' ? 'h6' : 'h4'}>
        {type === 'edit'
          ? 'This board is empty. Create a new column to get started.'
          : 'There are no Kanban boards currently available. Create a new board to get started...'}
      </Typography>
      <Box
        display="flex"
        justifyContent="start"
        sx={{ flexGrow: 1, mr: 1, mt: 2 }}
      >
        <FormModal
          triggerBtn={{
            type: 'normal',
            // @ts-ignore
            icon: type === 'edit' ? EditOutlinedIcon : AddIcon,
            text: type === 'edit' ? 'Edit Board' : 'Add New Board',
            color: 'secondary',
          }}
          title={type === 'edit' ? 'Edit Board' : 'Add New Board'}
        >
          <KanbanBoardForm
            isEditMode={type === 'edit'}
            currentData={currentData}
          />
        </FormModal>
      </Box>
    </div>
  );
};

export default KanbanBoardEmpty;
