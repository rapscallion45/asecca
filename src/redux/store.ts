import { configureStore, Action } from '@reduxjs/toolkit';
import { ThunkAction } from 'redux-thunk';
import logger from 'redux-logger';
import userPermissionReducer from './slices/userPermissionSlice';
import costsConfigReducer from './slices/costsConfigSlice';
import notificationsReducer from './slices/notificationsSlice';
import themeReducer from './slices/themeSlice';
import { collectionsKanbanSlice } from './slices/kanbanSlice';

/**
 * Configuration of the redux store to manage the global application state
 *
 * @author Carl Scrivener {@link https://github.com/rapscallion45 GitHub}
 * @since 0.0.0
 *
 * @see See [more info on Redux Store object](https://redux.js.org/api/store)
 *
 * @constant
 * @type {ToolkitStore<Object>}
 */
const store = configureStore({
  reducer: {
    userPermission: userPermissionReducer,
    costsConfig: costsConfigReducer,
    notifications: notificationsReducer,
    theme: themeReducer,
    collectionsKanban: collectionsKanbanSlice.reducer,
  },
  // @ts-ignore
  middleware:
    process.env.NODE_ENV === 'development'
      ? (getDefaultMiddleware) => getDefaultMiddleware().concat(logger)
      : (getDefaultMiddleware) => getDefaultMiddleware(),
});

/**
 * Shorthand application store dispatch type helper
 *
 * @author Carl Scrivener {@link https://github.com/rapscallion45 GitHub}
 * @since 0.0.0
 *
 * @typedef AppDispatch - application specific store dispatch type
 */
export type AppDispatch = typeof store.dispatch;

/**
 * Shorthand application state type helper
 *
 * @author Carl Scrivener {@link https://github.com/rapscallion45 GitHub}
 * @since 0.0.0
 *
 * @typedef AppState - application specific state object shape
 */
export type AppState = ReturnType<typeof store.getState>;

/**
 * Shorthand application Thunk Action type helper
 *
 * @author Carl Scrivener {@link https://github.com/rapscallion45 GitHub}
 * @since 0.0.0
 *
 * @typedef AppThunk - application specific Thunk action definition
 */
export type AppThunk = ThunkAction<void, AppState, unknown, Action<string>>;

export default store;
