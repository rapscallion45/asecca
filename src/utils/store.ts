/* redux store configuration placeholder */
const useStore = () => {};
export default useStore;
