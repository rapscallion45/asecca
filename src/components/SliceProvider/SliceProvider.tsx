import { FC, ReactNode, createContext, useContext } from 'react';
import { useSelector } from 'react-redux';
import { Slice } from '@reduxjs/toolkit';
import { AppState } from '@/redux/store';

/**
 * Default value for the slice context
 *
 * @author Carl Scrivener {@link https://github.com/rapscallion45 GitHub}
 * @since 0.0.8
 * @memberof SliceProvider
 *
 * @constant
 * @type {Slice<Object>}
 */
const SliceContext = createContext<Slice>({} as Slice);

/**
 * Slice Provider Props
 *
 * @author Carl Scrivener {@link https://github.com/rapscallion45 GitHub}
 * @since 0.0.8
 *
 * @typedef ISliceProviderProps
 * @prop {Slice} slice - context slice to provide down to child nodes
 * @prop {ReactNode} children - child nodes
 */
interface ISliceProviderProps {
  slice: Slice;
  children: ReactNode;
}

/**
 * Slice Provider
 *
 * Slice provider functional component wrapper for providing global state
 * managment functionality to child components
 *
 * @author Carl Scrivener {@link https://github.com/rapscallion45 GitHub}
 * @since 0.0.8
 *
 * @component
 * @param {ISliceProviderProps} props - component props
 * @returns {FC} - slice provider functional component
 */
const SliceProvider: FC<ISliceProviderProps> = (props) => {
  const { slice, children } = props;
  return (
    <SliceContext.Provider value={slice}>{children}</SliceContext.Provider>
  );
};

/**
 * Provider specific actions interface for passed slice
 *
 * @author Carl Scrivener {@link https://github.com/rapscallion45 GitHub}
 * @since 0.0.8
 * @memberof KanbanReduxSlice
 *
 * @see See [more info on Redux Slice creation](https://redux-toolkit.js.org/api/createSlice)
 *
 * @constant
 * @type {CaseReducerActions<any, string>}
 */
export const useSliceActions = () => useContext<Slice>(SliceContext).actions;

/**
 * Provider specific selector interface for passed slice
 *
 * @author Carl Scrivener {@link https://github.com/rapscallion45 GitHub}
 * @since 0.0.8
 * @memberof KanbanReduxSlice
 *
 * @see See [more info on Redux Slice creation](https://redux-toolkit.js.org/api/createSlice)
 *
 * @constant
 * @type {AppState}
 */
export const useSliceSelector = () => {
  const { name } = useContext<Slice>(SliceContext);
  return useSelector((state: AppState) => state[name as keyof AppState]);
};

export default SliceProvider;
