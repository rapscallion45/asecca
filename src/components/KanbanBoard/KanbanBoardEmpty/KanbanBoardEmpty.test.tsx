import React from 'react';
import { Provider } from 'react-redux';
import { render, screen, waitFor, act } from '@testing-library/react';
import renderer from 'react-test-renderer';
import {
  collectionsKanbanSlice,
  fetchByProjectId as fetchKanbanBoardByProjectId,
} from '@/redux/slices/collectionsKanbanSlice';
import store from '@/redux/store';
import KanbanBoardEmpty from './KanbanBoardEmpty';
import SliceProvider from '../../SliceProvider/SliceProvider';

/* default test query ID */
const query: string = '66135000001760012';

/* test params */
const testType: string = 'edit';

/**
 * Kanban Board Empty Unit Tests
 *
 * @author Carl Scrivener {@link https://github.com/rapscallion45 GitHub}
 * @since 0.0.22
 */
describe('Kanban Board Empty', () => {
  describe('Functions', () => {
    /** ensure mock function calls are cleared after test */
    afterEach(() => {
      jest.clearAllMocks();
    });

    it('Should render "Create a new column" text if edit mode enabled', async () => {
      /** Arrange */
      const editTestType: string = 'edit';
      act(() => {
        store.dispatch(
          fetchKanbanBoardByProjectId({
            projectId: query,
          })
        );
      });

      /** Act */
      render(
        <Provider store={store}>
          <SliceProvider slice={collectionsKanbanSlice}>
            <KanbanBoardEmpty type={editTestType} canEdit />
          </SliceProvider>
        </Provider>
      );

      /** Assert - menu button should be rendered and popup not displayed */
      await waitFor(() => {
        expect(
          screen.getByText(
            'This board is empty. Create a new column to get started.'
          )
        ).toBeInTheDocument();
      });
    });

    it('Should render "Add New Board" button if add type', async () => {
      /** Arrange */
      const editTestType: string = 'add';
      act(() => {
        store.dispatch(
          fetchKanbanBoardByProjectId({
            projectId: query,
          })
        );
      });

      /** Act */
      render(
        <Provider store={store}>
          <SliceProvider slice={collectionsKanbanSlice}>
            <KanbanBoardEmpty type={editTestType} />
          </SliceProvider>
        </Provider>
      );

      /** Assert - menu button should be rendered and popup not displayed */
      await waitFor(() => {
        expect(screen.getByText('Add New Board')).toBeInTheDocument();
      });
    });
  });

  describe('Error and Loading', () => {
    /** ensure mock function calls are cleared after test */
    afterEach(() => {
      jest.clearAllMocks();
    });

    it('Should render loading skeleton when loading prop set', async () => {
      /** Arrange */
      act(() => {
        store.dispatch(
          fetchKanbanBoardByProjectId({
            projectId: query,
          })
        );
      });

      /** Act */
      act(() => {
        render(
          <Provider store={store}>
            <SliceProvider slice={collectionsKanbanSlice}>
              <KanbanBoardEmpty type={testType} />
            </SliceProvider>
          </Provider>
        );
      });

      /** Assert - loading skeleton should be rendered while API fetches data */
      await waitFor(() => {
        expect(
          screen.getByTestId('kanban-board-loading-skeleton')
        ).toBeInTheDocument();
      });
    });
  });

  it('Renders correctly', async () => {
    /** Arrange */
    act(() => {
      store.dispatch(
        fetchKanbanBoardByProjectId({
          projectId: query,
        })
      );
    });

    /** perform snapshot test */
    const tree = renderer
      .create(
        <Provider store={store}>
          <SliceProvider slice={collectionsKanbanSlice}>
            <KanbanBoardEmpty type={testType} />
          </SliceProvider>
        </Provider>
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});
