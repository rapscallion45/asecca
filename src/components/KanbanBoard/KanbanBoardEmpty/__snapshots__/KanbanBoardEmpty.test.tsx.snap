// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Kanban Board Empty Renders correctly 1`] = `
<div
  className="MuiBox-root css-1xhj18k"
  data-testid="kanban-board-loading-skeleton"
>
  <div
    className="MuiBox-root css-1ay9vb9"
  >
    <span
      className="MuiSkeleton-root MuiSkeleton-text MuiSkeleton-pulse css-8n1xzc-MuiSkeleton-root"
      style={
        {
          "height": 50,
          "width": 190,
        }
      }
    />
    <span
      className="MuiSkeleton-root MuiSkeleton-rectangular MuiSkeleton-pulse css-nko6dn-MuiSkeleton-root"
      style={
        {
          "height": 100,
          "width": 270,
        }
      }
    />
    <span
      className="MuiSkeleton-root MuiSkeleton-rectangular MuiSkeleton-pulse css-nko6dn-MuiSkeleton-root"
      style={
        {
          "height": 100,
          "width": 270,
        }
      }
    />
  </div>
  <div
    className="MuiBox-root css-1ay9vb9"
  >
    <span
      className="MuiSkeleton-root MuiSkeleton-text MuiSkeleton-pulse css-8n1xzc-MuiSkeleton-root"
      style={
        {
          "height": 50,
          "width": 190,
        }
      }
    />
    <span
      className="MuiSkeleton-root MuiSkeleton-rectangular MuiSkeleton-pulse css-nko6dn-MuiSkeleton-root"
      style={
        {
          "height": 100,
          "width": 270,
        }
      }
    />
    <span
      className="MuiSkeleton-root MuiSkeleton-rectangular MuiSkeleton-pulse css-nko6dn-MuiSkeleton-root"
      style={
        {
          "height": 100,
          "width": 270,
        }
      }
    />
  </div>
  <div
    className="MuiBox-root css-1ay9vb9"
  >
    <span
      className="MuiSkeleton-root MuiSkeleton-text MuiSkeleton-pulse css-8n1xzc-MuiSkeleton-root"
      style={
        {
          "height": 50,
          "width": 190,
        }
      }
    />
    <span
      className="MuiSkeleton-root MuiSkeleton-rectangular MuiSkeleton-pulse css-nko6dn-MuiSkeleton-root"
      style={
        {
          "height": 100,
          "width": 270,
        }
      }
    />
    <span
      className="MuiSkeleton-root MuiSkeleton-rectangular MuiSkeleton-pulse css-nko6dn-MuiSkeleton-root"
      style={
        {
          "height": 100,
          "width": 270,
        }
      }
    />
  </div>
</div>
`;
